sources/engine/Xenko.Graphics/Direct3D/SwapChainGraphicsPresenter.Direct3D.cs

```diff
@@ -151,7 +154,26 @@ public override void Present()
         {
             try
             {
-                swapChain.Present((int)PresentInterval, PresentFlags.None);
+                if (isOccluded)
+                {
+                    var result = swapChain.Present((int)PresentInterval, PresentFlags.Test);
+                    if (result == Result.Ok)
+                    {
+                        isOccluded = false;
+                    }
+                    else
+                    {
+                        Thread.Sleep(1);
```

Comment:
What is the reason for sleeping here?

I'm worried `Thread.Sleep(1)` might cause some issues. There is no guarantee a thread will sleep for only 1 ms. In fact, it is more likely to sleep for much more than that (OS interrupt, context switching, etc.). If a sleep can't be avoided, better use `Thread.Yield()` or `Thread.Sleep(0)`.

Reply:
Use Thread.Yield() or Thread.Sleep(0). The CPU is still very exaggerated. Use Thread.Sleep(1) shouldn't have much impact, even if it sleeps 15 milliseconds. look this [https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/dxgi-status (DXGI_STATUS_OCCLUDED)](https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/dxgi-status)

Reply:
If Utilities.Sleep is more accurate, it can also be used.

Reply:
With this change, the game loop while minimized runs at a rate higher than refresh rate which doesn't seem expected ?
This is probably because rendering is running asynchronously to logic ?

Reply:
No ? As Kryptos said: sleep is unreliable, using it as is to try to produce constant frame times is unwise.
If you aren't too sure how to properly implement it, verify that you understand how the game and render loop are setup and find a couple of articles on the subjet, try searching for "programming constant framerate" or similar on google.

Reply:
When the window is minimized, the game screen cannot be rendered and the frame rate cannot be maintained. At this time only PressFlags.Test (not rendering) is waiting for recovery.
My understanding is that the game screen is now sleeping and waiting for recovery.

Reply:
I'm not sure I understood what you're saying, we're talking about the game loop here, we can't stop it while the window is minimized, right ?
We could implement a toggle to reduce the game's update rate or stop it entirely, we can safely ignore rendering but I'm pretty sure that by default *even while minimized* the expected behavior is that the game loop keeps running in the background.